substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -127,6 +187,227 @@ pub mod pallet {
+      let mut key = ValidatorSet { session: Session(new_index), network };
+      let mut joining =
+        Self::joining_validators(key).iter().map(|(id, _)| *id).collect::<BTreeSet<T::AccountId>>();
+
+      // get only the ones who is leaving because of insufficient bond.
+      let leaving = Self::deallocating_validators(key)
+        .iter()
+        .filter(|(_, _, leaving)| *leaving)
+        .map(|(id, _, _)| *id)
+        .collect::<Vec<_>>();
+
+      // get the current set
+      key = ValidatorSet { session: Self::current_session(), network };
+      let mut current = Self::validator_set(key)
+        .unwrap()
+        .participants
+        .iter()
+        .map(|(id, _)| *id)
+        .collect::<BTreeSet<T::AccountId>>();
+
+      // add new ones
+      // - We have to add new ones before filtering out the leaving ones.
+      // Since some of the leaving ones are still only in the joining set.
+      // These are the ones who allocated to a set and then
+      // immediately deallocated before getting into an active set when
+      // they are still only in the joining set. If they are to be removed from
+      // joining set as a result of their deallocation, we remove them here.
+      // We can't remove at the time of their public call, because we still
+      // need to deallocate the funds.
+      //
+      // - Alternatively we can just block the deallocation if you are not
+      // in an active set currently.
+      current.append(&mut joining);
+
+      // remove the validators who wanted to leave
+      current.retain(|id| !leaving.contains(id));
+
+      Vec::from_iter(current)
+    }
+
+    /// Makes a new validator set for the given session index with the given `validators`
+    /// as participants. All validators in `validators`, must be present either
+    /// in `JoiningValidators` or the current active validator set.
+    /// Does not fire the `Event::NewSet` for Serai network, since this expected to fired
+    /// prior to this call for a given session index.
+    pub fn start_session(new_index: u32, network: NetworkId, validators: Vec<T::AccountId>) {
+      // validator sets for index 0 is already set in the genesis build.
+      if new_index == 0 {
+        return;
+      }
+      let key = ValidatorSet { session: Session(new_index), network };
+
+      // get the bond of the validators, prepare participants.
+      let mut participants = Vec::new();
+      let mut joining = JoiningValidators::<T>::get(key);
+      let deallocating = DeallocatingValidators::<T>::get(key)
+        .iter()
+        .filter(|(_, _, leaving)| !leaving)
+        .map(|(id, amount, _)| (*id, *amount))
+        .collect::<Vec<_>>();
+      let prev =
+        Self::validator_set(ValidatorSet { session: Session(new_index - 1), network }).unwrap();
+
+      // append them all together
+      joining.extend(prev.participants);
+      let last_add_index = joining.len() - 1;
+      joining.extend(deallocating);
```

Comment:
Why is `deallocating` merged into `joining`?

Reply:
all 3 vectors merged into 1 to not repeat [following](https://github.com/akildemir/serai/blob/baaf96f59c06b1157e75059d275cc69de5625f1a/substrate/validator-sets/pallet/src/lib.rs#L348) block of code.